### PR TITLE
Interpolator

### PR DESCRIPTION
**Pull request by yisraelU**
I believe there are still more use cases to figure out

### CODE REVIEW

File: modules/codegen/src/smithy4s/codegen/Interpolator.scala [hunk 1]
```diff
@@ -0,0 +1,33 @@
+package smithy4s.codegen
+
+import cats.implicits._
+object Interpolator {
+  class Interpolator(val sc: StringContext) extends AnyVal {
```
Comment (Baccata) [inline modules/codegen/src/smithy4s/codegen/Interpolator.scala:5]: make this an implicit class (and rename it to `LineInterpolator`)

File: modules/codegen/src/smithy4s/codegen/Interpolator.scala [hunk 2]
```diff
@@ -0,0 +1,33 @@
+package smithy4s.codegen
+
+import cats.implicits._
+object Interpolator {
```
Comment (Baccata) [inline modules/codegen/src/smithy4s/codegen/Interpolator.scala:4]: rename this to `LineSyntax`

File: build.sbt [hunk 3]
```diff
@@ -348,8 +348,9 @@ lazy val codegen = projectMatrix
       Dependencies.Weaver.cats.value % Test
     ),
     testFrameworks += new TestFramework("weaver.framework.CatsEffect"),
-    scalacOptions := scalacOptions.value
-      .filterNot(Seq("-Ywarn-value-discard", "-Wvalue-discard").contains)
+ scalacOptions := scalacOptions.value
+      .filterNot(Seq("-Ywarn-value-discard", "-Wvalue-discard","-Xfatal-warnings").contains)
```
Comment (Baccata) [inline build.sbt:352]: todo : revert before finalising the PR

File: modules/codegen/src/smithy4s/codegen/Interpolator.scala [hunk 4]
```diff
@@ -0,0 +1,33 @@
...
+  class Interpolator(val sc: StringContext) extends AnyVal {
+    def render(renderables: Renderable.WithValue[_]*): RenderLine = {
```
Comment (Baccata) [inline modules/codegen/src/smithy4s/codegen/Interpolator.scala:6]: let's rename this method to `line`

File: modules/codegen/src/smithy4s/codegen/Renderable.scala [hunk 5]
```diff
@@ -64,7 +107,26 @@ object Renderable {
   }
 }
 
+// Models
+case class RenderLine(imports: Set[String], line: String) {
```
Comment (Baccata) [inline modules/codegen/src/smithy4s/codegen/Renderable.scala:111]: Let's rename this to `Line`

File: modules/codegen/src/smithy4s/codegen/Renderable.scala [hunk 6]
```diff
@@ -64,7 +107,26 @@ object Renderable {
...
+  def tupled = (imports, line)
+  def modify(f: String => String) = RenderLine(imports, f(line))
+  def toRenderResult = RenderResult(imports, List(line))
+}
+object RenderLine {
```
Comment (Baccata) [inline modules/codegen/src/smithy4s/codegen/Renderable.scala:116]: A `Line` should have a `Renderable` instance

File: modules/example/src/smithy4s/example/FooService.scala [hunk 7]
```diff
@@ -5,7 +5,7 @@ import smithy4s.syntax._
 trait FooServiceGen[F[_, _, _, _, _]] {
   self =>
 
-  def getFoo() : F[Unit, Nothing, GetFooOutput, Nothing, Nothing]
+  def getFoo() : F[PrimitiveType(Unit),Ref(smithy4s.example,GetFooOutput),List()]
```
Comment (Baccata) [inline modules/example/src/smithy4s/example/FooService.scala:8]: Something is obviously off there, the `toString` is being called on `Type` somewhere it shouldn't be

File: modules/codegen/src/smithy4s/codegen/Renderable.scala [hunk 8]
```diff
@@ -33,13 +33,56 @@ object Renderable {
 
   implicit val stringRenderable: Renderable[String] = (s: String) =>
     RenderResult(Set.empty, List(s))
-  implicit def listRenderable[A](implicit
-      A: Renderable[A]
-  ): Renderable[List[A]] = { (l: List[A]) =>
+  def typeRenderable(ns: String): Renderable[Type] = new Renderable[Type] {
```
Comment (Baccata) [inline modules/codegen/src/smithy4s/codegen/Renderable.scala:36]: I think we could make this non-parameterised, and filter out the imports that could be removed "at the end of the world"

File: modules/codegen/src/smithy4s/codegen/Renderer.scala [hunk 9]
```diff
@@ -338,11 +330,11 @@ private[codegen] class Renderer(compilationUnit: CompilationUnit) { self =>
...
       hints: List[Hint]
   ): RenderResult = {
-    val decl = s"case class $name(${renderArgs(fields)})"
-    val imports = fields.foldMap(_.tpe.imports) ++ syntaxImport
+    val decl = render"case class $name(${renderArgs(fields)})"
+    val imports = decl.imports ++ syntaxImport
     lines(
       if (hints.contains(Hint.Error)) {
-        block(s"$decl extends Throwable") {
+        block(s"${decl.line} extends Throwable") {
```
Comment (Baccata) [inline modules/codegen/src/smithy4s/codegen/Renderer.scala:337]: as you correctly inferred, let's make `block` take a `Line`, and have a method overload that'd take a `String`

File: modules/codegen/src/smithy4s/codegen/Renderer.scala [hunk 10]
```diff
@@ -594,6 +592,27 @@ private[codegen] class Renderer(compilationUnit: CompilationUnit) { self =>
 
     def renderAlgParams =
       s"$renderInput, $renderError, $renderOutput, $renderStreamedInput, $renderStreamedOutput"
+
+    def renderAlgParamsWithImports(genName: String) = {
```
Comment (Baccata) [inline modules/codegen/src/smithy4s/codegen/Renderer.scala:596]: the implementation of this should look like this : 

``` 
import op._ 
line"$input, $errors, $output, $streamedInput, $streamedOutput"
```

**Comment by Baccata** [comment]
After giving this some thoughts, we should clean the names/terminology. Here's a list of actions : 

- [ ] `RenderLine` => `Line`
- [ ]  `RenderResult` => `Lines` (plural) 
- [ ]  `def render` interpolator would become `def line`
- [ ] `Renderable` => `ToLines` 
- [ ]   We should create a `ToLine` expressing the same thing as `ToLines` but for a single line
- [ ]  We currently have `Renderable.WithValue`. Let's make it more generic, and go with : 

```scala 
case class WithValue[TC[_], A](typeclass: TC[A], value: A) 

object WithValue {
   implicit def to[TC[_], A](value: A)(implicit instance: TC[A]): WithValue[TC, A] =
      WithValue(value, instance)
}
```

This way, we should be able to use `WithValue` for both `ToLines` and `ToLine`

- [ ]  Let's have this non-orphan typeclass instance : `implicit def lineToLines[A : ToLine] : ToLines[A]`
- [ ]  Let's move all possible instances from `ToLines` to `ToLine` 
- [ ]  Let's make a `ToLine` instance for `Type` and `Primitive`
- [ ]  Let's remove `importsAndRender` plainly and simply

File: modules/example/src/smithy4s/example/FooService.scala (hunk 7 above)
Comment (yisraelU) [reply modules/example/src/smithy4s/example/FooService.scala:8]: yeah forgot to checkin .

File: modules/codegen/src/smithy4s/codegen/Renderer.scala (hunk 10 above)
Comment (yisraelU) [reply modules/codegen/src/smithy4s/codegen/Renderer.scala:596]: problem I have is to return a Line , I must include the imports for Error , which require the Service Name . The problem is the Operation doesn't have access to the Service name, hence I was passing it in .
Should I leave out the Error imports and add them in via addImports function (which kind of returns us back to this ad hoc style..)

Comment (Baccata) [reply modules/codegen/src/smithy4s/codegen/Renderer.scala:596]: The operations (and their error coproducts) are rendered within the service's companion object, so as long as we `import MyService._`  in the file of of the service, it should be okay really

**Comment by Baccata** [comment]
Closing in favour of https://github.com/disneystreaming/smithy4s/pull/209

**Pull request closed without merging**